pkg/querier/queryrange/marshal.go

```diff
@@ -109,135 +41,106 @@ func WriteSeriesResponseProtobuf(version loghttp.Version, r logproto.SeriesRespo
+				Status: "success",
+				Data: queryrangebase.PrometheusData{
+					ResultType: loghttp.ResultTypeMatrix,
+					Result:     sampleStream,
 				},
 			},
+			Statistics: result.Statistics,
 		}, nil
 	case promql.Scalar:
 		sampleStream, err := queryrangebase.FromValue(data)
 		if err != nil {
 			return nil, err
 		}
 
-		return &QueryResponse{
-			Response: &QueryResponse_Prom{
-				Prom: &LokiPromResponse{
-					Response: &queryrangebase.PrometheusResponse{
-						Status: "success",
-						Data: queryrangebase.PrometheusData{
-							ResultType: loghttp.ResultTypeScalar,
-							Result:     sampleStream,
-						},
-					},
-					Statistics: result.Statistics,
+		return &LokiPromResponse{
+			Response: &queryrangebase.PrometheusResponse{
+				Status: "success",
+				Data: queryrangebase.PrometheusData{
+					ResultType: loghttp.ResultTypeScalar,
+					Result:     sampleStream,
 				},
 			},
+			Statistics: result.Statistics,
 		}, nil
 	case logqlmodel.Streams:
-		return &QueryResponse{
-			Response: &QueryResponse_Streams{
-				// Note: we are omitting the Version here because the Protobuf already defines a schema.
-				Streams: &LokiResponse{
-					Direction: params.Direction(),
-					Limit:     params.Limit(),
-					Data: LokiData{
-						ResultType: loghttp.ResultTypeStream,
-						Result:     data,
-					},
-					Status:     "success",
-					Statistics: result.Statistics,
-				},
+		return &LokiResponse{
+			Direction: params.Direction(),
+			Limit:     params.Limit(),
+			Data: LokiData{
+				ResultType: loghttp.ResultTypeStream,
+				Result:     data,
 			},
+			Status:     "success",
+			Statistics: result.Statistics,
 		}, nil
 	case sketch.TopKMatrix:
 		sk, err := data.ToProto()
-		if err != nil {
-			return nil, err
-		}
-		return &QueryResponse{
-			Response: &QueryResponse_TopkSketches{
-				TopkSketches: &TopKSketchesResponse{Response: sk},
-			},
-		}, nil
+		return &TopKSketchesResponse{Response: sk}, err
 	case sketch.QuantileSketchMatrix:
-		return &QueryResponse{
-			Response: &QueryResponse_QuantileSketches{
-				QuantileSketches: &QuantileSketchResponse{Response: data.ToProto()},
-			},
-		}, nil
+		return &QuantileSketchResponse{Response: data.ToProto()}, nil
 	}
 
 	return nil, fmt.Errorf("unsupported data type: %t", result.Data)
 }
+
+func QueryResponseWrap(res queryrangebase.Response) (*QueryResponse, error) {
+	p := &QueryResponse{}
+
+	switch response := res.(type) {
+	case *LokiPromResponse:
+		p.Response = &QueryResponse_Prom{response}
+	case *LokiResponse:
+		p.Response = &QueryResponse_Streams{response}
+	case *LokiSeriesResponse:
+		p.Response = &QueryResponse_Series{response}
+	case *MergedSeriesResponseView:
+		mat, err := response.Materialize()
+		if err != nil {
+			return p, err
+		}
+		p.Response = &QueryResponse_Series{mat}
+	case *LokiLabelNamesResponse:
+		p.Response = &QueryResponse_Labels{response}
+	case *IndexStatsResponse:
+		// TODO: *queryrange.IndexStatsResponse
```

Comment:
TODO

Reply:
I won't adress it in this PR though. Should I remove it?